Add page for websocket (#537)
vegaprotocol/documentation: docs/api/websocket.md

---
title: Websocket Streams
sidebar_label: Websocket Streams
---
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

Vega's API contains websocket endpoints that offer real-time updates to changes in state of the Vega network, allowing subscriptions to events such as per-market trades or changes to a party's position.

As Vega is a blockchain time moves in discrete blocks and so updates will appear as blocks are executed.

## Authentication and Ratelimiting

API tokens are not required to access the API as they are all public. TLS is supported on all websocket endpoints but note that whether it is enabled on a particular data node is a choice made by the node operator.

> Websocket connections are ratelimited by a maximum allowed number of subscriptions per IP address. The default maximum is set to 250 connections but note that this value may differ between data node operators.

Reviewer: ```suggestion
Websocket connections are rate limited by a maximum allowed number of subscriptions per IP address. The default maximum is set to 250 connections but note that this value may differ between data node operators.
```

Subscription to a websocket endpoint happens when the connection is opened and unsubscription occurs when the connection is closed. It is not necessary to send a request through the websocket to initiate the subscription or to prove liviness of the connection.

## Subscribing using the WebSockets API

<Tabs>
<TabItem value="bash" label="Bash">

```bash
curl -L -X GET 'https://api.n07.testnet.vega.xyz/api/v2/stream/ledger/movements'
```

</TabItem>

<TabItem value="py" label="Python">

```py
import rel
import json
import websocket

url = "wss://api.n07.testnet.vega.xyz/api/v2/stream/ledger/movements"

def on_message(ws, message):
    print(json.loads(message))

ws = websocket.WebSocketApp(
    url=url,
    on_message=on_message,
)

ws.run_forever(dispatcher=rel, reconnect=5)
rel.signal(2, rel.abort)
rel.dispatch()
```

</TabItem>

<TabItem value="js" label="Node">

```js
var WebSocketClient = require('websocket').client;

> url = "wss://api.n07.testnet.vega.xyz/api/v2/stream/ledger/movements"

Reviewer: ```suggestion
var url = "wss://api.n07.testnet.vega.xyz/api/v2/stream/ledger/movements"
```

      "balances": [
        {
          "account": {
            "assetId": "fBTC",
            "type": 4,
            "owner": "ee023a39d8c76b7c32b235f4620fb50cae2af3068368979c41eaeb519ce8d3fd"
          },
          "balance": "30"
        }
      ]
    }
  }
}
```

All enum values are sent as their integer values and not their string representation. This is to reduce the amount of data in each packet and to help the performance.

## Snapshot Data

Some of the Websocket endpoints will send a snapshot of the current state of data when a connection is first made. This allows for an application to build an initial state creating context for subsequent updates. The snapshot data will be sent in batches after which subsequent messages will only be updates to the snapshot state.

As an example, when streaming orders the current state of the orderbook will be sent first

```json
{
    "result": {
        "snapshot": {
            "orders": [],
            "lastPage: true,
        }
    }
}
```

The last batch of snapshot data will have `lastPage` set to `true` after which the stream will switch to sending updates to the orderbook

```json
{
    "result": {
        "updates": {

>             "orders": [],

Reviewer: ```suggestion
            "orders": []
```

As above
